### PR TITLE
Allow users to control redirect behavior.

### PR DESCRIPTION
**Pull request by brianquinlan**

### CODE REVIEW

File: src/CUPHTTPClientDelegate.m [hunk 1]
```diff
@@ -83,6 +116,41 @@ - (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)task
...
+  Dart_CObject ctype = MessageTypeToCObject(RedirectMessage);
+  Dart_CObject credirect = NSObjectToCObject(redirect);
+  Dart_CObject* message_carray[] = { &ctype, &credirect };
+
+  Dart_CObject message_cobj;
+  message_cobj.type = Dart_CObject_kArray;
+  message_cobj.value.as_array.length = 2;
+  message_cobj.value.as_array.values = message_carray;
+
+  [redirect.lock lock];  // After this line, any attempt to acquire the lock will wait.
```
Comment (aam) [inline src/CUPHTTPClientDelegate.m:141]: do you have to unlock this lock?

File: src/CUPHTTPClientDelegate.m [hunk 2]
```diff
@@ -83,6 +116,41 @@ - (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)task
...
+  const bool success = Dart_PostCObject_DL(config.sendPort, &message_cobj);
+  if (!success) {
+    os_log_error(OS_LOG_DEFAULT, "Dart_PostCObject_DL failed.");
+    completionHandler(nil);
+    return;
+  }
+  [redirect.lock lock];  // Wait until CUPHTTPRedirect is done.
```
Comment (aam) [inline src/CUPHTTPClientDelegate.m:148]: Is this as intended? It's not clear why lock acquired few lines above need to be acquired again.

File: src/CUPHTTPClientDelegate.m (hunk 1 above)
Comment (aam) [reply src/CUPHTTPClientDelegate.m:141]: perhaps worth adding a comment pointing to the place where the lock gets release

**Review by aam — APPROVED** [review]

File: src/CUPHTTPClientDelegate.m (hunk 2 above)
Comment (brianquinlan) [reply src/CUPHTTPClientDelegate.m]: I added a bigger comment and a pointer to the overall strategy.

File: src/CUPHTTPClientDelegate.m (hunk 1 above)
Comment (brianquinlan) [reply src/CUPHTTPClientDelegate.m:141]: I added a larger comment to the next lock call.

**Pull request merged**